File: go/core/builder.go
package core

import (
	"context"
	"encoding/json"
	"fmt"
	"slices"
)

type definition struct {
	plugins []Plugin
	order   []int
}

func (d definition) copy() definition {
	return definition{slices.Clone(d.plugins), slices.Clone(d.order)}
}

// Builders are mutable, confined to one goroutine. Built Grammars are immutable.
type GrammarBuilder struct {
	runtime    *Runtime
	definition definition
}

func (r *Runtime) Builder() *GrammarBuilder { return &GrammarBuilder{runtime: r} }
func (b *GrammarBuilder) Add(plugin *Plugin) error {
	if plugin == nil {
		return fmt.Errorf("nil plugin")
	}
	for _, p := range b.definition.plugins {
		if p.identity == plugin.identity {
			return &BuildError{Code: "duplicate", Element: plugin.name}
		}
	}
	if plugin.providerRuntime != nil && plugin.providerRuntime != b.runtime {
		return fmt.Errorf("text provider belongs to another runtime")
	}
	order := slices.Clone(b.definition.order)
	for _, rule := range plugin.rules {
		if rule.runtime != b.runtime {
			return fmt.Errorf("rule belongs to another runtime")
		}
		if slices.Contains(order, rule.index) {
			return &BuildError{Code: "duplicate", Element: rule.name}
		}
		order = append(order, rule.index)
	}
	b.definition.plugins = append(b.definition.plugins, *plugin)
	b.definition.order = order
	return nil
}
func (b *GrammarBuilder) Remove(plugin *Plugin) error {
	if plugin == nil {
		return fmt.Errorf("nil plugin")
	}
	index := slices.IndexFunc(b.definition.plugins, func(p Plugin) bool { return p.identity == plugin.identity })
	if index < 0 {
		return &BuildError{Code: "missing", Element: plugin.name}
	}
	for _, rule := range b.definition.plugins[index].rules {
		b.definition.order = slices.DeleteFunc(b.definition.order, func(i int) bool { return i == rule.index })
	}
	b.definition.plugins = slices.Delete(b.definition.plugins, index, index+1)
	return nil
}
func (b *GrammarBuilder) Before(rule, anchor *Rule) error {
	if rule == nil || anchor == nil || rule.runtime != b.runtime || anchor.runtime != b.runtime {
		return fmt.Errorf("rules must belong to this runtime")
	}
	if rule.phase != anchor.phase {
		return fmt.Errorf("rules must belong to the same phase")
	}
	from, to := slices.Index(b.definition.order, rule.index), slices.Index(b.definition.order, anchor.index)
	if from < 0 {
		return &BuildError{Code: "missing", Element: rule.name}
	}
	if to < 0 {
		return &BuildError{Code: "missing", Element: anchor.name}
	}
	if from != to {
		b.definition.order = slices.Delete(b.definition.order, from, from+1)
		if from < to {
			to--
		}
		b.definition.order = slices.Insert(b.definition.order, to, rule.index)
	}
	return nil
}
func (b *GrammarBuilder) Build(ctx context.Context) (*Grammar, error) {
	snapshot := b.definition.copy()
	recipe, err := json.Marshal(snapshot.composition())
	if err != nil {
		return nil, err
	}
	var grammar *Grammar
	// Reserve across the whole pool: every accepted grammar must fit on every
	// worker when first parsed, regardless of which worker built it.
	if err := b.runtime.reserveGrammar(); err != nil {
		return nil, err
	}
	defer func() {
		if grammar == nil {
			b.runtime.grammarCount.Add(-1)
		}
	}()
	err = b.runtime.withWorker(ctx, func(w *worker) error {
		result, err := w.build(ctx, recipe)
		if err != nil {
			return err
		}
		grammar = b.runtime.newGrammar(snapshot, recipe, result.Description)
		grammar.state.initialized.Store(true)
		w.grammars[grammar.state] = result.Handle
		return nil
	})
	return grammar, err
}

File: typescript/src/core/parser/builder.ts
import type { Owner, Snapshot, Rule } from "./types.js";
import type { Plugin } from "./plugin.js";
import type { PluginGroup } from "../definitions/index.js";
import { data, owned } from "./identity.js";
import { GrammarBuildError } from "./errors.js";

export class GrammarBuilder {
  #runtime: Owner;
  #plugins: Snapshot["plugins"];
  #order: number[];
  constructor(runtime: Owner, snapshot: Snapshot = { plugins: [], order: [] }) {
    this.#runtime = runtime;
    this.#plugins = [...snapshot.plugins];
    this.#order = [...snapshot.order];
  }
  add(plugin: Plugin) {
    const state = data(plugin, "plugin");
    if (this.#plugins.some((p) => p.symbol === state.symbol))
      throw new GrammarBuildError({
        code: "duplicate",
        element: state.name ?? "<anonymous plugin>",
      });
    for (const provider of state.text)
      if (provider.runtime !== this.#runtime)
        throw new TypeError("Text provider belongs to a different Runtime");
    const used = new Set(this.#order);
    for (const rule of state.rules) {
      const { index } = owned(rule, "rule", this.#runtime);
      if (used.has(index))
        throw new GrammarBuildError({
          code: "duplicate",
          element: rule.name ?? "<anonymous rule>",
        });
      used.add(index);
    }
    this.#plugins.push(state);
    this.#order.push(...state.rules.map((r) => data(r, "rule").index));
    return this;
  }
  remove(plugin: Plugin) {
    const state = data(plugin, "plugin");
    const index = this.#plugins.findIndex((p) => p.symbol === state.symbol);
    if (index < 0)
      throw new GrammarBuildError({
        code: "missing",
        element: state.name ?? "<anonymous plugin>",
      });
    const removed = new Set(
      this.#plugins[index].rules.map((r) => data(r, "rule").index),
    );
    this.#plugins.splice(index, 1);
    this.#order = this.#order.filter((index) => !removed.has(index));
    return this;
  }
  before(rule: Rule<"inline">, anchor: Rule<"inline">): this;
  before(rule: Rule<"block">, anchor: Rule<"block">): this;
  before(rule: Rule<"text">, anchor: Rule<"text">): this;
  before(rule: Rule, anchor: Rule) {
    const first = owned(rule, "rule", this.#runtime),
      next = owned(anchor, "rule", this.#runtime);
    if (first.phase !== next.phase)
      throw new TypeError("Rules must belong to the same phase");
    const from = this.#order.indexOf(first.index),
      to = this.#order.indexOf(next.index);
    if (from < 0 || to < 0)
      throw new GrammarBuildError({
        code: "missing",
        element: (from < 0 ? rule : anchor).name ?? "<anonymous rule>",
      });
    if (from !== to) {
      const [entry] = this.#order.splice(from, 1);
      this.#order.splice(from < to ? to - 1 : to, 0, entry);
    }
    return this;
  }
  build() {
    return this.#runtime.build({
      plugins: [...this.#plugins],
      order: [...this.#order],
    });
  }
}
export function composition(snapshot: Snapshot) {
  const groups: { parent: number | null; name: string }[] = [],
    indices = new Map<PluginGroup, number>();
  function visit(group: PluginGroup | null): number | null {
    if (!group) return null;
    if (indices.has(group)) return indices.get(group)!;
    const parent = visit(group.parent),
      index = groups.length;
    groups.push({ parent, name: group.name });
    indices.set(group, index);
    return index;
  }
  const plugins = snapshot.plugins.map((p) => ({
    group: visit(p.group),
    name: p.name,
    rules: p.rules.map((rule) => data(rule, "rule").index),
    text: p.text.map((provider) => provider.index),
  }));
  return { groups, plugins, order: snapshot.order };
}
